### ISSUE TITLE
restore is not the inverse of dump

### ISSUE THREAD

**Issue by Mulugruntz** [edgedb/edgedb#2433]
- EdgeDB Version: Docker image: edgedb/edgedb:1-beta1
- OS Version:

Steps to Reproduce:

1. `edgedb dump backup --all --format=dir`
2. `edgedb restore -v --allow-non-empty --all backup`

Result:
```
edgedb error: error applying init file "backup/init.edgeql": failed statement "# DESCRIBE ROLES\nCREATE SUPERUSER ROLE `myuser` { SET password_hash := \'SCRAM-SHA-256$4096:xxxxxxxxxxx=\'};": ERROR: SchemaError: role 'myuser' is already present in the schema
```

I would expect that if I put `--allow-non-empty`, especially with `--all`, I would be allowed to do such a thing.
Is it as intended?

I feel like there should be a way to cleanly "start fresh from a full backup".

**Comment by Mulugruntz** [comment]
By the way, if I do:
```
$> edgedb drop-role myuser
$> edgedb restore -v --allow-non-empty --all backup
edgedb error: ERROR: AuthenticationError: authentication failed
```

Hmmm.... maybe I would need to use the `edgedb` role for that, instead? Not sure how to do it...

**Comment by elprans** [comment]
>  Is it as intended?

Generally yes.  Unlike Postgres, EdgeDB will not let you restore schema objects that already exist in the target database or instance.

> Hmmm.... maybe I would need to use the `edgedb` role for that, instead? Not sure how to do it...

The `edgedb` role is the default, so you can either omit `-e EDGEDB_USER` or specify `-e EDGEDB_USER=edgedb` explicitly when starting up a fresh server in a container.

That said, it seems we need to consider emitting `ROLE` objects with `IF NOT EXISTS` in dumps.

**Comment by Mulugruntz** [comment]
yes, agreed with the last statement :-).

In my dump scripts, I'll try to amend the file to put `IF NOT EXISTS` in the meantime.

**Comment by Mulugruntz** [comment]
Erm... there's no `IF NOT EXISTS` statement to the role creation, is there?
https://www.edgedb.com/docs/edgeql/admin/roles#create-role

**Comment by Mulugruntz** [comment]
After commenting out the `CREATE SUPERUSER ROLE`, I still have an error message
```
edgedb error: restoring database "edgedb": Error initiating restore protocol: ERROR: SchemaDefinitionError: specified migration parent is not the most recent migration, expected 'm14zgmivx7oosifsy4oht376mmbmlb3bhwue5dbj7xj464uwqogjvq'
```

What would you advise?

**Comment by elprans** [comment]
> What would you advise?

Use the `edgedb` role.

**Comment by Mulugruntz** [comment]
With the `edgedb` role.
```
edgedb error: ERROR: AuthenticationError: invalid SCRAM verifier for user 'edgedb'
```

**Comment by elprans** [comment]
You need to use `edgedb` when you are starting the server container on a fresh data volume for the first time to initialize it with the supplied password.  Switching the role afterwards does not reset its password.

**Comment by Mulugruntz** [comment]
Not sure when I can do that.
Because my docker-compose will start both services and will create everything.

Maybe I misunderstand something, but as of now, dump/restore doesn't seem usable, does it?
I haven't managed to make it work, so far.

I was expecting something like `pg_dump`. https://www.postgresql.org/docs/13/app-pgdump.html
`pg_dump -Fc > filename` and `pg_restore filename`

I think I'm gonna try a different strategy:
Backup:

1. stop the server
2. tar.gz the Filesystem, storing it on host.
3. start the server

Restore:
1. stop the server
2. mount a volume with the tar.gz, extract the tar.gz
3. start the server

**Comment by elprans** [comment]
> Maybe I misunderstand something, but as of now, dump/restore doesn't seem usable, does it?

The issue is that you are using a custom `ROLE` instead of the default (`edgedb`).  When you start a docker container for the first time on an empty data volume, it configures the authentication for that data volume (with values from `EDGEDB_USER` and `EDGEDB_PASSWORD`).  If you've used a non-default role, and then attempted to restore a full backup it'd crash with the "role already present" error.  

(we'll make a change to teach the dump to be tolerant of overlapping roles in beta 2)

The solution for your issue is as follows: 

1. Make a dump from the original data volume
2. Stop the server.
3. Make a new, empty directory for the data volume.
4. Remove `EDGEDB_USER` from your `docker-compose.yml`.
5. Start the container.
6. Restore the dump.

**Comment by elprans** [comment]
Alternatively, don't use `--all` and restore databases one by one.

**Comment by elprans** [comment]
@Mulugruntz, can you share your `docker-compose.yml` please?  Also, did you try the steps I outlined above?

**Comment by Mulugruntz** [comment]
I tried to do what you advised me (and some other stuff as well).
But I wasn't sure about steps 3 & 4...

And to script the dump/restore, without destroying the whole environment, from my understanding, that doesn't seem practical?

**docker-compose.yml**
<details>

```yaml
version: '3'
services:
  tempbot:
    container_name: tempbot
    build: .
    ports:
      - "4242:8000" # The webserver
    volumes:
    - ./logs:/var/log
    links:
    - edgedb
    depends_on:
    - edgedb
  edgedb:
    container_name: edgedb
    image: edgedb/edgedb:1-beta1
    environment:
      - EDGEDB_USER=user
      - EDGEDB_PASSWORD=password
#      - EDGEDB_DEBUG_PGSERVER=1
    ports:
    - "5656:5656" # EdgeQL over binary protocol
    - "8889:8889" # EdgeQL over HTTP
    - "8888:8888" # GraphQL over HTTP
    volumes:
    - edgedb-db:/var/lib/edgedb/data
volumes:
  edgedb-db: {}
```

</details>

**Comment by elprans** [comment]
> But I wasn't sure about steps 3 & 4...

Hm, maybe I'm not explaining it right.  Here, in terms of the steps you outlined above:

> So, when I try:
> 
> 1. Create a clean DB with docker-compose up (includes initial migration)
> 2. Populate the db a bit.
> 3. Run `backup_db.sh`
> 4. `docker exec -i edgedb edgedb -c "select count(ImportedFile);"` -> `{7}`
> 5. Run `docker container kill edgedb ; docker container rm edgedb ; docker volume rm temp-bot-async_edgedb-db`

**5.1. Remove `EDGEDB_USER` from `docker-compose`.** 

> 6. Create a clean DB with docker-compose up (includes initial migration)
> 7. Run `restore_db.sh`
> 8. `docker exec -i edgedb edgedb -c "select count(ImportedFile);"` -> `{0}`

`backup_db.sh` should run `edgedb dump --all` and `restore_db.sh` should run `edgedb restore --all`

**Comment by elprans** [comment]
> (includes initial migration)

Oh, and please don't run the migration before restoring from dump, otherwise you'll run into "already exists" errors on other schema objects.

**Comment by Mulugruntz** [comment]
Okay, I'm gonna try all that.
Could I suggest a new `--wipe-clean` or `--force` (or another name) for `edgedb restore`?
This would make the whole think so much more usable. 

Therefore, just a `edgedb -u edgedb restore --wipe-clean` would make it work all the time (or any other one-liner that would guarantee it). If the EdgeDB team would see value in that, we could open a new ticket.

**Comment by Mulugruntz** [comment]
Still no luck.

**backup_db.sh**

<details>

```bash
#!/usr/bin/env bash

# exit on error
set -e

get_abs_filename() {
  # $1 : relative filename
  python -c "from pathlib import Path;print(f'''{Path('$1').resolve()}''')"
}

this_dir=$(get_abs_filename ${0%/*})
spinner="$this_dir/spinner.sh"

DOCKER_CNAME=edgedb

TS=$(date -u +%Y%m%d_%H%M%S%Z)
BACKUP_FILENAME="backup_${TS}.edgedb.dump.tar.gz"
BACKUP_FILEPATTERN="backup_*.edgedb.dump.tar.gz"

BACKUP_LOCAL_DIR=$(get_abs_filename "$this_dir/../backup")
mkdir -p "$BACKUP_LOCAL_DIR"

REMOTE_DATADIR=$(docker exec -it "$DOCKER_CNAME" /usr/bin/env bash -c 'echo $EDGEDB_DATADIR')
REMOTE_DATADIR="${REMOTE_DATADIR//[$'\t\r\n ']}"
REMOTE_BACKUP_MOUNT="/backup"

echo "Creating the dump..."
$spinner docker exec edgedb edgedb dump backup --all --format=dir

echo "Compressing the dump..."
$spinner docker exec edgedb tar -czvf "$BACKUP_FILENAME" backup

echo "Copying the dump file to the host machine..."
$spinner docker cp "edgedb:$BACKUP_FILENAME" "$BACKUP_LOCAL_DIR/"

echo "Cleanup..."
$spinner docker exec edgedb rm -rf backup "$BACKUP_FILENAME"

echo "Done!"

BACKUP_FILELIST=$(get_abs_filename "$BACKUP_LOCAL_DIR/$BACKUP_FILEPATTERN")
echo "Created dump file $BACKUP_FILENAME"
ls -lah $BACKUP_FILELIST
```

</details>

**restore_db.sh**

<details>

```bash
#!/usr/bin/env bash

# exit on error
set -e

if [ -z "$1" ]; then
  echo "No backup file provided."
  exit 1
fi

get_abs_filename() {
  # $1 : relative filename
  python -c "from pathlib import Path;print(f'''{Path('$1').resolve()}''')"
}

DOCKER_CNAME=edgedb
EDGE_VOLUME=$(docker inspect -f '{{ (index .Mounts 0).Name }}' "$DOCKER_CNAME")

function shutdown() {
  echo "Starting the database..."
  $spinner docker container start "$DOCKER_CNAME"
}
trap shutdown EXIT

this_dir=$(get_abs_filename ${0%/*})
root_dir=$(get_abs_filename "$this_dir/../..")
spinner="$this_dir/spinner.sh"

BACKUP_FILEPATH=$(get_abs_filename "$1")
BACKUP_FILENAME=$(basename "$BACKUP_FILEPATH")
BACKUP_DIRNAME=$(dirname "$BACKUP_FILEPATH")

REMOTE_DATADIR=$(docker exec -it "$DOCKER_CNAME" /usr/bin/env bash -c 'echo $EDGEDB_DATADIR')
REMOTE_DATADIR="${REMOTE_DATADIR//[$'\t\r\n ']}"
REMOTE_BACKUP_MOUNT="/backup"

echo "Stopping the database..."
$spinner docker container stop "$DOCKER_CNAME"
echo "Removing the database..."
$spinner docker container rm "$DOCKER_CNAME"
echo "Removing the volume..."
docker volume rm "$EDGE_VOLUME"

echo "Creating a new container with edgedb role..."
docker-compose -f "$root_dir/docker-compose.yml" --env-file "$this_dir/.env.restore" up -d --no-deps --force-recreate --build "$DOCKER_CNAME"

echo "Restoring the dump..."
$spinner docker exec edgedb mkdir -p "$REMOTE_BACKUP_MOUNT/backup"
echo "   Copying the dump file to the container..."
$spinner docker cp "$BACKUP_FILEPATH" "edgedb:$REMOTE_BACKUP_MOUNT/$BACKUP_FILENAME"
echo "   Extracting the dump..."
$spinner docker exec edgedb tar -C "$REMOTE_BACKUP_MOUNT/backup" -xzvf "$REMOTE_BACKUP_MOUNT/$BACKUP_FILENAME" --strip 1
echo "   EdgeDB restoring..."
$spinner docker exec edgedb edgedb restore -v --allow-non-empty --all "$REMOTE_BACKUP_MOUNT/backup"
RESULT=$?

echo "Cleanup..."
$spinner docker exec edgedb rm -rf "$REMOTE_BACKUP_MOUNT"

echo "Creating a new container with user role..."
docker-compose -f "$root_dir/docker-compose.yml" --env-file "$root_dir/.env" up -d --no-deps --force-recreate --build "$DOCKER_CNAME"


if [ $RESULT -eq 0 ]; then
    echo "Done!"
else
    echo "Error!"
fi
```

</details>

**bin/.env.restore**

<details>

```bash
EDGEDB_USER="edgedb"
EDGEDB_PASSWORD=""
```

</details>

If I run the script, this is the relevant failing part:
```
   EdgeDB restoring...
edgedb error: Cannot assign requested address (os error 99)
```

And if I run it myself in the container:
```bash
root@166e28f41fd0:/# edgedb restore -v --allow-non-empty --all /backup/backup
edgedb error: ERROR: AuthenticationError: invalid SCRAM verifier for user 'edgedb'
```
```bash
root@166e28f41fd0:/# env
HOSTNAME=166e28f41fd0
PWD=/
EDGEDB_USER=edgedb
EDGEDB_PASSWORD=
HOME=/root
LANG=en_US.utf8
EDGEDB_DATADIR=/var/lib/edgedb/data
GOSU_VERSION=1.11
TERM=xterm
SHLVL=1
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
VERSION=1-beta1
_=/usr/bin/env
```

I think the os error 99 is because when `docker-compose` gives back the control, the DB is not yet ready to init.

**Comment by Mulugruntz** [comment]
It works!
I put a dummy password in `.env.restore`:
```bash
EDGEDB_USER="edgedb"
EDGEDB_PASSWORD="edgepassword"
```

And I reran it.

```bash
...
...
   EdgeDB restoring...
[2021-04-09T08:47:23Z INFO  edgedb::commands::restore] Schema applied in 17.4524872s
[2021-04-09T08:47:23Z INFO  edgedb::commands::restore] Blocks sent in 327.8379ms
[2021-04-09T08:47:35Z INFO  edgedb::commands::restore] Complete in 29.8203546s
Cleanup...
Creating a new container with user role...
Recreating edgedb ... done
Done!
Starting the database...
edgedb
(temp-bot-async) ⋊> ~/P/f/temp-bot-async on CBA-2_populate_database ⨯   docker exec -i edgedb edgedb -c "select count(ImportedFile);"                                 17:47:52
6
```

I just need to monitor the docker logs on the 2nd (user) container creation, wait on ready and make the cleanup a bit more serious.

But thanks for your dedication and help!

(my  suggestion on  `--wipe-clean` or `--force` still stands 😄 )

**Comment by Mulugruntz** [comment]
I created this gist to show the full solution (cleanup is now okayish).
https://gist.github.com/Mulugruntz/a06cfd81d3953816c5a4b79fb32066ed

If you have any suggestion, I'm all ears. But at least, it works!
